### PR TITLE
Support for compiling with MSYS

### PR DESCRIPTION
**Pull request by rafalcieslak**
This pull request includes support for compiling NFD using the MSYS/MinGW environment.

### CODE REVIEW

File: src/nfd_win.cpp [hunk 1]
```diff
@@ -360,14 +367,15 @@ nfdresult_t NFD_OpenDialog( const char *filterList,
     HRESULT result = ::CoInitializeEx(NULL,
                                       ::COINIT_APARTMENTTHREADED |
                                       ::COINIT_DISABLE_OLE1DDE );
+
```
Comment (rafalcieslak) [inline src/nfd_win.cpp]: GCC considers it an error when a goto jump crosses variable initialisation, thus fileOpenDialog has to be initialized a bit earlier.

File: src/nfd_win.cpp [hunk 2]
```diff
@@ -360,14 +367,15 @@ nfdresult_t NFD_OpenDialog( const char *filterList,
     HRESULT result = ::CoInitializeEx(NULL,
                                       ::COINIT_APARTMENTTHREADED |
                                       ::COINIT_DISABLE_OLE1DDE );
+                                      
+    ::IFileOpenDialog *fileOpenDialog(NULL);
```
Comment (rafalcieslak) [inline src/nfd_win.cpp]: GCC considers it an error when a goto jump crosses variable initialisation, thus fileOpenDialog has to be initialized a bit earlier.

**Comment by DenisKolodin** [comment]
@rafalcieslak Thank you for this patch! It helps me a lot!

**Comment by DenisKolodin** [comment]
We are forced to maintain own copy of this lib there: https://github.com/saurvs/nfd-rs
But it has no special edits which won't be useful here. Last change we need to use this library as a submodule is this PR (mingw support). Because mingw is a main toolchain of Rust under Windows.

@mlabbe Could you revise and merge it, please?

**Comment by mlabbe** [comment]
I am in the process of deprecating SCons and using Premake-generated build files to build Native File Dialog, removing the most common complaint about the library.  Unlike most Premake-based projects, I am checking in the generated scripts so people won't need Premake.

If you check out the devel branch, the build/gmake_windows directory contains a makefile.  I am not a mingw user; can you confirm you can use it to build the devel branch?

I will be merging devel to master shortly.  It would be great to have Mingw support in the mainline code when I do.

Thanks for your help.

**Comment by DenisKolodin** [comment]
I haven't seen `devel` branch before. It contains a lot of changes with a build system.

I tried to check it, but `premake5` builds with mingw, but don't work with msys (I don't know why there is a difference). `premake4` for msys is available.

I built 5th for msys here https://github.com/premake/premake-core/issues/559. Now I can check `devel` branch and I'll report about it. The problem is quite deeper than I expected )

**Comment by mlabbe** [comment]
Ideally, the `build/gmake_windows` dir contains a generated makefile that would work with msys and you wouldn't have to run Premake yourself.  It was generated with `premake5 --os=windows gmake`

The devel branch's nfd_win.cpp is EXACTLY the same as master's.  Only the build system has changed.

Again, since I don't use these compilers, I'm relying on your assistance to support them.  Thanks for taking a look!

**Comment by DenisKolodin** [comment]
I checked `build/gmake_windows` build script for mingw. If I add this PR's patch then `nfd` builds, but tests don't and fails with errors:
```c
==== Building nfd (release_x64) ====
==== Building test_opendialog (release_x64) ====
Linking test_opendialog
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0x599): undefined reference to `__imp_CoInitializeEx'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0x5d0): undefined reference to `__imp_CoCreateInstance'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0x67b): undefined reference to `__imp_CoTaskMemFree'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0x6c4): undefined reference to `__imp_CoUninitialize'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0x755): undefined reference to `__imp_CoInitializeEx'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0x78c): undefined reference to `__imp_CoCreateInstance'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0x7ad): undefined reference to `__imp_CoUninitialize'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0xb49): undefined reference to `__imp_CoInitializeEx'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0xb80): undefined reference to `__imp_CoCreateInstance'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0xba1): undefined reference to `__imp_CoUninitialize'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.text+0xc31): undefined reference to `__imp_CoTaskMemFree'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.rdata$.refptr.CLSID_FileSaveDialog[.refptr.CLSID_FileSaveDialog]+0x0): undefined reference to `CLSID_FileSaveDialog'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.rdata$.refptr.IID_IFileSaveDialog[.refptr.IID_IFileSaveDialog]+0x0): undefined reference to `IID_IFileSaveDialog'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.rdata$.refptr.CLSID_FileOpenDialog[.refptr.CLSID_FileOpenDialog]+0x0): undefined reference to `CLSID_FileOpenDialog'
../lib/Release/x64/nfd.lib(nfd_win.o):nfd_win.cpp:(.rdata$.refptr.IID_IFileOpenDialog[.refptr.IID_IFileOpenDialog]+0x0): undefined reference to `IID_IFileOpenDialog'
collect2.exe: error: ld returned 1 exit status
```

To fix link errors I've added the following to test's makefiles (manually):
```make
LIBS += ../lib/Release/x64/nfd.lib -lole32 -luuid
```

I tried to fix `LIBS` in `premake5.lua`, but I take an error when I'm trying to generate build script:
```
Error: ./nfd/build/premake5.lua:30: attempt to call global 'symbols' (a nil value)
```

In total:
1) This PR works to build library for msys/mingw with premake script
2) To build tests it's necessary to add `ole32` and `uuid` dependencies

> Note about the second, I think VS links that dependencies automatically.

**Comment by mlabbe** [comment]
Okay, great.  That's a start.  I still need:
1. instructions on how to install your compiler, so I can test this myself
2. a working premake file

How to build premake so you don't get an error:
https://github.com/mlabbe/nativefiledialog/blob/master/docs/build.md

A few rules on how to submit pull requests:
https://github.com/mlabbe/nativefiledialog/blob/master/docs/submitting_pull_requests.md

I won't be able to just locally add some libs and trust that it works.  I need a working pull request targeting the devel branch that I can test with instructions.  As mentioned, I don't use the mingw compiler, so I need to know how you get it and set it up.

**Comment by DenisKolodin** [comment]
Ok, I'll do it and send you instructions. Also, I'll fix `premake` script.
Should I send a new PR which fixes it to `devel` branch or we'll continue to work here?

**Comment by mlabbe** [comment]
Great!  Just do another PR.  We can link to it here if anyone wants to follow along.

**Comment by DenisKolodin** [comment]
I've made another PR #19 with premake fixes.

**Comment by mlabbe** [comment]
PR #19 was merged into the devel branch today.

**Pull request closed without merging**